File: backend/refs/matlab/_aux_fun/kalman_filter_covid.m
function  output = kalman_filter_covid(y,x,F,A,H,Q,R_base,kappa_seq,settings)

% This function runs the Kalman Filter given parameters. Following Hamilton's (1994) notation
% 
%   State Equation
%   xi_{t+1} = F xi_{t} + v_{t+1}
%
%   Measurement Eq
%   y_t = A' x_t + H' xi_{t} + w_t
%
%   Error covariance
%   Q = E[v_t v_t']
%   R = E[w_t w_t']
%
% INPUTS
% y = T x n_y matrix of n_y measurment time series observations
% x = T x n_x matrix of n_x exogenous variables. This includes constants and lags.
% F = n_s x n_s matrix transition for states 
% A = n_x x n_y matrix for exogenous variables
% H = n_s x n_y matrix for states variables
% Q = n_s x n_s, var-cov matrix for innovations in the state eq.
% R = n_y x n_y, var-cov matrix for innovations in the measurement eq.
% settings: object that contains several options



[T_y,n_y] = size(y);
[T_x,n_x] = size(x);
n_s = size(F,1);

if (T_y ~= T_x); error('y and x have different number of observations'); end;

T = T_y;

% create placeholders for the inferred variables

% xi_{t|t}, contemporanous prediction
xi_mat_c = zeros(T,n_s);
% xi_{t+1|t}, one step ahead prediction
xi_mat = zeros(T,n_s);

% xi_{t|T}, smoothed
xi_smoothed_mat = zeros(T,n_s);

% P_{t|t}
P_mat_c = zeros(T,n_s,n_s);
% P_{t+1|t}
P_mat = zeros(T,n_s,n_s);
% P_{t|T}, smoothed
P_smoothed_mat = zeros(T,n_s,n_s);

% J_{t}
J_mat = zeros(T,n_s,n_s);


loglik_vec = zeros(T,1);


% initial value of the state
try
    xi_0 = settings.xi_0';
catch
    xi_0 = zeros(n_s,1)';
end

% initial value of the precision
try
   P_0   = settings.P_0;
catch
    P_0  = solve_ricatti(F,Q);
end

for t=1:T
    if t == 1
        P_tt = P_0 ;
        xi_tt = xi_0';
    else
        P_tt  = squeeze(P_mat_c(t-1,:,:));
        xi_tt = xi_mat_c(t-1,:)';
    end
    R = R_base*(kappa_seq(t)^2); %this accounts for the time-varying matrix

    %update state one step ahead forecast
    xi_mat(t,:) = (F*xi_tt)';

    %update one step ahead forecast prediction error
    P_ttm1 =  F*P_tt*F'+Q;
    P_mat(t,:,:) = P_ttm1;
    
    % Prediction error
    pred_error = y(t,:)'-(A')*(x(t,:)')-(H')*(xi_mat(t,:)');
    HPHR = (H')*P_ttm1 *H+R;

    % Given this compute the likelihood of the obs using eq 13.4.1 in Hamilton
    loglik_vec(t) = -(n_y/2)*log(2*pi)-0.5*log(det(HPHR))...
        -0.5*pred_error'*(HPHR\pred_error);

    % Contemporaneous prediction
    xi_mat_c(t,:) = xi_mat(t,:)'+...
        (P_ttm1*H)*(HPHR\pred_error);

     % Contemporaneous prediction error:
    P_mat_c(t,:,:) = P_ttm1-P_ttm1*H*(HPHR\(H'*P_ttm1));
end

% given this, now run the Kalman smoother to obtain smoothed state
% estimates.

% CHECK THIS AGAINST HLW

if settings.smoother == 1

states_smoother = settings.states_smoother;
n_smoother = length(states_smoother);
% start the smoother
xi_smoothed_mat(T,states_smoother) =  xi_mat_c(T,states_smoother);
P_smoothed_mat(T,states_smoother,states_smoother) = squeeze(P_mat_c(T,states_smoother,states_smoother));
for t = T-1:-1:1
    P_tt = squeeze(P_mat_c(t,states_smoother,states_smoother));
    P_tp1t = squeeze(P_mat(t+1,states_smoother,states_smoother)); 
    %J_t =  P_tt*F(states_smoother,states_smoother)'*(P_tp1t\eye(n_smoother));
    J_t =  P_tt*F(states_smoother,states_smoother)'*pinv(P_tp1t);
    J_mat(t,states_smoother,states_smoother) = J_t;
    xi_tt = (xi_mat_c(t,states_smoother)');
    xt_tp1t = xi_mat(t+1,states_smoother)';
    xi_tT = xi_tt + J_t*((xi_smoothed_mat(t+1,states_smoother)')-xt_tp1t);
    P_tT = P_tt + J_t*(squeeze(P_smoothed_mat(t+1,states_smoother,states_smoother))-P_tp1t)*J_t';
    xi_smoothed_mat(t,states_smoother) = xi_tT;
    P_smoothed_mat(t,states_smoother,states_smoother) = P_tT;
end

end


output.state_contemporaneous = xi_mat_c;
output.state_contemporaneous_pred = P_mat_c;
output.state_one_ahead = xi_mat;
output.state_one_ahead_pred = P_mat;
output.state_smoothed = xi_smoothed_mat;
output.state_smoothed_pred = P_smoothed_mat;
output.loglik_vec = loglik_vec;

File: backend/pipeline/_aux_fun/kalman_filter_covid.py
from __future__ import annotations

from typing import Any

import numpy as np
from scipy.linalg import solve_discrete_lyapunov


def _get(obj: Any, name: str) -> Any:
    if isinstance(obj, dict):
        return obj[name]
    return getattr(obj, name)


def solve_ricatti(F: np.ndarray, Q: np.ndarray) -> np.ndarray:
    """Fallback used when P_0 is not supplied."""
    return solve_discrete_lyapunov(F, Q)


def kalman_filter_covid(
    y: np.ndarray,
    x: np.ndarray,
    F: np.ndarray,
    A: np.ndarray,
    H: np.ndarray,
    Q: np.ndarray,
    R_base: np.ndarray,
    kappa_seq: np.ndarray,
    settings,
) -> dict[str, np.ndarray]:
    """
    Run the Kalman filter.

    State equation
        xi_{t+1} = F xi_t + v_{t+1}

    Measurement equation
        y_t = A' x_t + H' xi_t + w_t
    """
    y = np.asarray(y, dtype=float)
    x = np.asarray(x, dtype=float)
    F = np.asarray(F, dtype=float)
    A = np.asarray(A, dtype=float)
    H = np.asarray(H, dtype=float)
    Q = np.asarray(Q, dtype=float)
    R_base = np.asarray(R_base, dtype=float)
    kappa_seq = np.asarray(kappa_seq, dtype=float).reshape(-1)

    T_y, n_y = y.shape
    T_x, _ = x.shape
    n_s = F.shape[0]

    if T_y != T_x:
        raise ValueError("y and x have different number of observations")

    T = T_y

    # xi_{t|t}, contemporaneous prediction
    xi_mat_c = np.zeros((T, n_s))
    # xi_{t+1|t}, one-step-ahead prediction
    xi_mat = np.zeros((T, n_s))
    # xi_{t|T}, smoothed
    xi_smoothed_mat = np.zeros((T, n_s))

    # P_{t|t}
    P_mat_c = np.zeros((T, n_s, n_s))
    # P_{t+1|t}
    P_mat = np.zeros((T, n_s, n_s))
    # P_{t|T}, smoothed
    P_smoothed_mat = np.zeros((T, n_s, n_s))

    # J_t
    J_mat = np.zeros((T, n_s, n_s))
    loglik_vec = np.zeros(T)

    # Initial value of the state
    try:
        xi_0 = np.asarray(_get(settings, "xi_0"), dtype=float).reshape(-1)
    except Exception:
        xi_0 = np.zeros(n_s)

    # Initial value of the covariance
    try:
        P_0 = np.asarray(_get(settings, "P_0"), dtype=float)
    except Exception:
        P_0 = solve_ricatti(F, Q)

    for t in range(T):
        if t == 0:
            P_tt = P_0
            xi_tt = xi_0
        else:
            P_tt = P_mat_c[t - 1, :, :]
            xi_tt = xi_mat_c[t - 1, :]

        # Time-varying measurement covariance
        R = R_base * (kappa_seq[t] ** 2)

        # One-step-ahead prediction
        xi_pred = F @ xi_tt
        xi_mat[t, :] = xi_pred

        P_ttm1 = F @ P_tt @ F.T + Q
        P_mat[t, :, :] = P_ttm1

        # Prediction error
        pred_error = y[t, :] - (A.T @ x[t, :]) - (H.T @ xi_pred)
        HPHR = H.T @ P_ttm1 @ H + R

        # Likelihood contribution of observation t
        sign, logdet = np.linalg.slogdet(HPHR)
        if sign <= 0:
            raise np.linalg.LinAlgError("Measurement covariance HPHR is not positive definite.")

        HPHR_inv_pred_error = np.linalg.solve(HPHR, pred_error)
        loglik_vec[t] = (
            -(n_y / 2.0) * np.log(2.0 * np.pi)
            - 0.5 * logdet
            - 0.5 * pred_error.T @ HPHR_inv_pred_error
        )

        # Contemporaneous prediction
        xi_upd = xi_pred + (P_ttm1 @ H) @ HPHR_inv_pred_error
        xi_mat_c[t, :] = xi_upd

        # Contemporaneous prediction error
        P_upd = P_ttm1 - P_ttm1 @ H @ np.linalg.solve(HPHR, H.T @ P_ttm1)
        P_mat_c[t, :, :] = P_upd

    # Run the smoother if requested
    if int(_get(settings, "smoother")) == 1:
        states_smoother = np.asarray(_get(settings, "states_smoother"), dtype=int)

        # MATLAB indices are 1-based; convert when passed through directly.
        if states_smoother.min() >= 1:
            states_smoother = states_smoother - 1

        xi_smoothed_mat[T - 1, states_smoother] = xi_mat_c[T - 1, states_smoother]
        P_smoothed_mat[np.ix_([T - 1], states_smoother, states_smoother)] = (
            P_mat_c[T - 1][np.ix_(states_smoother, states_smoother)]
        )

        for t in range(T - 2, -1, -1):
            P_tt = P_mat_c[t][np.ix_(states_smoother, states_smoother)]
            P_tp1t = P_mat[t + 1][np.ix_(states_smoother, states_smoother)]

            J_t = P_tt @ F[np.ix_(states_smoother, states_smoother)].T @ np.linalg.pinv(P_tp1t)
            J_mat[t][np.ix_(states_smoother, states_smoother)] = J_t

            xi_tt = xi_mat_c[t, states_smoother]
            xi_tp1_pred = xi_mat[t + 1, states_smoother]
            xi_tp1_smooth = xi_smoothed_mat[t + 1, states_smoother]

            xi_tT = xi_tt + J_t @ (xi_tp1_smooth - xi_tp1_pred)
            P_tT = (
                P_tt
                + J_t
                @ (
                    P_smoothed_mat[t + 1][np.ix_(states_smoother, states_smoother)]
                    - P_tp1t
                )
                @ J_t.T
            )

            xi_smoothed_mat[t, states_smoother] = xi_tT
            P_smoothed_mat[t][np.ix_(states_smoother, states_smoother)] = P_tT

    return {
        "state_contemporaneous": xi_mat_c,
        "state_contemporaneous_pred": P_mat_c,
        "state_one_ahead": xi_mat,
        "state_one_ahead_pred": P_mat,
        "state_smoothed": xi_smoothed_mat,
        "state_smoothed_pred": P_smoothed_mat,
        "J_mat": J_mat,
        "loglik_vec": loglik_vec,
    }